pkg/operator/ceph/object/objectstore.go

```diff
@@ -916,11 +916,14 @@ func adjustZoneDefaultPools(zone map[string]interface{}, spec cephv1.ObjectShare
 		"user_uid_pool":   ".meta.users.uid",
 		"otp_pool":        ".otp",
 		"notif_pool":      ".log.notif",
+		"topics_pool":     ".meta.topics",  // introduced in Ceph v19
+		"account_pool":    ".meta.account", // introduced in Ceph v19
+		"group_pool":      ".meta.group",   // introduced in Ceph v19
 	}
 	for pool, nsSuffix := range zonePoolNSSuffix {
 		// replace rgw internal index pools with namespaced metadata pool
 		namespacedPool := defaultMetaPool + nsSuffix
-		prev, err := setObjProperty(zone, namespacedPool, pool)
+		prev, err := setObjPropertyIfExists(zone, namespacedPool, pool)
```

Comment:
An alternative to the "IfExists" approach, is to check the ceph version. Something like this:
```
if cephVersion.IsAtLeast(cephver.CephVersion{Major: 19, Minor: 0, Extra: 0}) {
  // add new properties
}
```
But if we can just check if the property exists, that may certainly keep things simpler in the long term when other properties are added. 

During an upgraded cluster (e.g. from v18 --> v19), the new properties will be found to exist and then we'll add them? Just wanted to confirm if you already tested that scenario.

Reply:
I think that Ceph version check approach is better for code readability. But, as you said, with "IfExists" approach we can just add any new pools and don't care about ceph version. 
I personally more like "IfExists" approach because it makes RGW responsible for pools structure and rook just tries to apply namespaces to all existing pools. 
I can also suggest iterating through all root string fields ending with `_pool`  and log warning if pool is unknown. Or try to apply some namespace anyway. What do you think?


I've upgaded my local rook-ceph from `v18.something` to `v19.1.1-0`. All daemon pods been restared and using v19 images, but when i run `radosgw-admin zone get` i don't see any new pools in json after update:
```json
{
    "id": "e522c205-dbd2-43e1-8ef8-a81a55d44ef1",
    "name": "pp-store",
    "domain_root": "meta-pool:pp-store.meta.root",
    "control_pool": "meta-pool:pp-store.control",
    "gc_pool": "meta-pool:pp-store.log.gc",
    "lc_pool": "meta-pool:pp-store.log.lc",
    "log_pool": "meta-pool:pp-store.log",
    "intent_log_pool": "meta-pool:pp-store.log.intent",
    "usage_log_pool": "meta-pool:pp-store.log.usage",
    "roles_pool": "meta-pool:pp-store.meta.roles",
    "reshard_pool": "meta-pool:pp-store.log.reshard",
    "user_keys_pool": "meta-pool:pp-store.meta.users.keys",
    "user_email_pool": "meta-pool:pp-store.meta.users.email",
    "user_swift_pool": "meta-pool:pp-store.meta.users.swift",
    "user_uid_pool": "meta-pool:pp-store.meta.users.uid",
    "otp_pool": "meta-pool:pp-store.otp",
    "system_key": {
        "access_key": "",
        "secret_key": ""
    },
    "placement_pools": [...],
    "realm_id": "ceb20ae5-3e69-40c4-af9f-fc9bab7904db",
    "notif_pool": "meta-pool:pp-store.log.notif"
}
```

Reply:
> I can also suggest iterating through all root string fields ending with _pool and log warning if pool is unknown. Or try to apply some namespace anyway. What do you think?

That sounds great to log a warning if we find an extra pool, until we are more confident with that approach.

> I've upgaded my local rook-ceph from v18.something to v19.1.1-0. All daemon pods been restared and using v19 images, but when i run radosgw-admin zone get i don't see any new pools in json after update:

This is what I worried about, the schema may not be updated until some zone update is triggered. So that means we may need to use the approach to check the ceph version instead of the "IfExists" approach.

Reply:
> That sounds great to log a warning if we find an extra pool, until we are more confident with that approach.

I really like this idea as well. 

-----

Conceptually, I much prefer the `IfExists` approach, but it does seem like we have an edge condition on upgrade we need figure out how to deal with properly. 

If we upgrade from v17/v18 -> v19, my expectation would be that RGW (upon upgrade) is going to create a new pool before we Rook has a chance to see if a new pool needs configured. I'm wondering what bad would happen if the RGW did that and started putting stuff into that pool before Rook has a chance to configure it. But it also seems like my expectation is not what happens IRL, which is also concerning.

I think it's best to ask @cbodley what his recommendation is for handling the upgrade case when new fields are introduced to the zone config. How does RGW operate on upgrade when the zone spec is missing configs for default pools? Does RGW create the new default pools and use them implicitly, or does it just not use the new default pools? Is there a good (preferably idempotent) way of checking to see if the zone config is missing fields it needs after an RGW upgrade? Are there other things around this we're missing?

I think the best-case scenario for Rook would be if RGW doesn't init new default pools on upgrade and we have a recommendation for detecting what pools RGW wants to be configured that we can use on an if-exists basis.

Reply:
> I've upgaded my local rook-ceph from `v18.something` to `v19.1.1-0`. All daemon pods been restared and using v19 images, but when i run `radosgw-admin zone get` i don't see any new pools in json after update:

@arttor that makes it sound like the `radosgw-admin` binary was still v18

Reply:
@BlaineEXE the upgrade case is handled by rgw during deserialization of `struct RGWZoneParams`: https://github.com/ceph/ceph/blob/squid/src/rgw/driver/rados/rgw_zone.h#L258-L265

if it finds an earlier version of the encoding (ie written by a reef rgw), it fills in missing pool names with their default values. after decode, rgw may go on to create and write into those pools

if rook wants to prevent this, it would have to use an upgraded `radosgw-admin zone set` command to update those pool names before any other radosgws/radosgw-admin commands get upgraded

reef rgws would still be able to read that updated encoding and, as long as they don't overwrite RGWZoneParams with their old encoding version, they'd see the updated pool names after upgrading to squid

Reply:
> I'm wondering what bad would happen if the RGW did that and started putting stuff into that pool before Rook has a chance to configure it.

if rook changes pool names after rgw has already created and written to the default pool names, that would essentially look like data loss

Reply:
it is probably something wrong with my setup. even when i try to create a new rgw, i don't see new pools in zone json. RGW is using this image [`quay.io/ceph/ceph@sha256:096f9e9a97636092e354ecc1a3fd411579d3c6adbbe1a195ad58ea84ad910de9
`](https://quay.io/repository/ceph/ceph/manifest/sha256:096f9e9a97636092e354ecc1a3fd411579d3c6adbbe1a195ad58ea84ad910de9)

Reply:
oh, yea. You are correct, i was using old `radosgw-admin` 🤦 . After toolbox update i can see new pools. 
For existing rgw with shared pools, new pools are using `<instance_name>.rgw.meta` pool:
```json
{
    "notif_pool": "meta-pool:pp-store.log.notif", // existing is using shared metadata pool "meta-pool"
    "topics_pool": "pp-store.rgw.meta:topics", // new pool using default "pp-store.rgw.meta"
    "account_pool": "pp-store.rgw.meta:accounts", 
    "group_pool": "pp-store.rgw.meta:groups",
...
}
```

Reply:
> if rook wants to prevent this, it would have to use an upgraded `radosgw-admin zone set` command to update those pool names before any other radosgws/radosgw-admin commands get upgraded

@BlaineEXE how can we achive this? It would be very easy if we could add new pools with old `radosgw-admin`. Or it is ok to ask users to upgrade rook to the latest version before upgrading Ceph? If it is ok, then we can just always force set new pools to zone and it should work.

Reply:
Since we know about the pools, can we attempt to always set them regardless of the Ceph version? That way, they will be ready on any upgrade.

Reply:
> that makes it sound like the radosgw-admin binary was still v18

That's a good point, and this is accurate. Because Ceph v19 is still pre-release, Rook is using the latest v18 image for its base image. That means that the Ceph CLI and all tools a re v18 tools, including radosgw-admin.